File: app/src/main/java/io/opentakserver/opentakicu/TakDataPackageImporter.java
package io.opentakserver.opentakicu;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.util.Log;
import android.util.Xml;

import androidx.preference.PreferenceManager;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringReader;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.KeyStore;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManagerFactory;

/**
 * Imports an ATAK / TAK Server Data Package (.zip) into OpenTAK ICU preferences.
 *
 * The zip is expected to contain:
 * - one {@code *.pref} XML file with a {@code cot_streams} preference block
 * - one or more {@code *.p12} files referenced from the .pref by {@code caLocation0}
 *   ({@code trust_store_certificate}) and {@code certificateLocation0} ({@code client_certificate})
 *
 * Used from both {@code ATAKPreferencesFragment} (settings UI) and {@code MainActivity}
 * (share / view intent from a file manager).
 */
public class TakDataPackageImporter {
    private static final String LOGTAG = "TakImporter";

    /** Result of an import attempt. */
    public static class Result {
        public boolean success;
        /** User-friendly summary or error message. */
        public String message;
        /** Server host (if parsed). null on failure. */
        public String host;
        /** Server port (if parsed). 0 on failure. */
        public int port;
        /** true if connectString0 protocol == ssl/https/tls. */
        public boolean ssl;
        /** Absolute path to the trust-store .p12 we extracted, if any. */
        public String trustStorePath;
        public String trustStorePassword;
        /** Absolute path to the client .p12 we extracted, if any. */
        public String clientCertPath;
        public String clientCertPassword;
    }

    public static class ConnectionTestResult {
        public boolean tcpOk;
        public boolean tlsOk;
        public String details;
    }

    /**
     * Extract the zip, parse the .pref, apply matching keys to the default SharedPreferences,
     * and return a {@link Result} describing what was imported.
     */
    public static Result importFromUri(Context ctx, Uri zipUri) {
        Result res = new Result();
        File filesDir = ctx.getFilesDir();
        Map<String, File> extracted = new HashMap<>();
        String prefXml = null;

        try (InputStream is = ctx.getContentResolver().openInputStream(zipUri);
             ZipInputStream zis = new ZipInputStream(is)) {
            ZipEntry entry;
            byte[] buf = new byte[4096];
            while ((entry = zis.getNextEntry()) != null) {
                if (entry.isDirectory()) {
                    zis.closeEntry();
                    continue;
                }
                String name = entry.getName().replace('\\', '/');
                String baseName = name.substring(name.lastIndexOf('/') + 1);
                String lower = baseName.toLowerCase();

                if (lower.endsWith(".pref")) {
                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
                    int n;
                    while ((n = zis.read(buf)) > 0) baos.write(buf, 0, n);
                    prefXml = baos.toString(StandardCharsets.UTF_8.name());
                } else if (lower.endsWith(".p12") || lower.endsWith(".pem") || lower.endsWith(".jks")) {
                    File dest = new File(filesDir, baseName);
                    try (FileOutputStream fos = new FileOutputStream(dest)) {
                        int n;
                        while ((n = zis.read(buf)) > 0) fos.write(buf, 0, n);
                    }
                    extracted.put(baseName, dest);
                    Log.d(LOGTAG, "Extracted " + baseName + " -> " + dest.getAbsolutePath());
                }
                zis.closeEntry();
            }
        } catch (Exception e) {
            res.success = false;
            res.message = "Could not read zip: " + e.getMessage();
            return res;
        }

        if (prefXml == null) {
            res.success = false;
            res.message = "No .pref file found inside the zip.";
            return res;
        }

        Map<String, String> values;
        try {
            values = parseCotStreamsPref(prefXml);
        } catch (Exception e) {
            res.success = false;
            res.message = "Could not parse .pref: " + e.getMessage();
            return res;
        }

        String connectString = values.get("connectString0");
        if (connectString == null) {
            res.success = false;
            res.message = "The .pref file is missing a connectString.";
            return res;
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(ctx);
        SharedPreferences.Editor editor = prefs.edit();
        StringBuilder summary = new StringBuilder();

        String[] parts = connectString.split(":");
        if (parts.length >= 1) {
            res.host = parts[0];
            editor.putString("atak_address", res.host);
            summary.append("Server: ").append(res.host).append('\n');
        }
        if (parts.length >= 2) {
            try {
                res.port = Integer.parseInt(parts[1].trim());
            } catch (NumberFormatException ignored) {}
            editor.putString("atak_port", parts[1]);
            summary.append("Port: ").append(parts[1]).append('\n');
        }
        if (parts.length >= 3) {
            String proto = parts[2].toLowerCase();
            res.ssl = proto.equals("ssl") || proto.equals("https") || proto.equals("tls");
            editor.putBoolean("atak_ssl", res.ssl);
            summary.append("Protocol: ").append(parts[2]).append('\n');
        }

        if ("true".equalsIgnoreCase(values.get("enabled0"))) {
            editor.putBoolean("send_cot", true);
            summary.append("CoT streaming: enabled\n");
        }

        String description = values.get("description0");
        if (description != null) {
            summary.insert(0, "Imported: " + description + "\n\n");
        }

        String caLoc = values.get("caLocation0");
        if (caLoc != null) {
            String basename = caLoc.replace('\\', '/');
            basename = basename.substring(basename.lastIndexOf('/') + 1);
            File f = extracted.get(basename);
            if (f != null) {
                res.trustStorePath = f.getAbsolutePath();
                editor.putString("trust_store_certificate", res.trustStorePath);
                summary.append("Trust store: ").append(basename).append('\n');
            }
        }
        if (values.containsKey("caPassword0")) {
            res.trustStorePassword = values.get("caPassword0");
            editor.putString("trust_store_cert_password", res.trustStorePassword);
        }

        String certLoc = values.get("certificateLocation0");
        if (certLoc != null) {
            String basename = certLoc.replace('\\', '/');
            basename = basename.substring(basename.lastIndexOf('/') + 1);
            File f = extracted.get(basename);
            if (f != null) {
                res.clientCertPath = f.getAbsolutePath();
                editor.putString("client_certificate", res.clientCertPath);
                summary.append("Client cert: ").append(basename).append('\n');
            }
        }
        if (values.containsKey("clientPassword0")) {
            res.clientCertPassword = values.get("clientPassword0");
            editor.putString("client_cert_password", res.clientCertPassword);
        }

        editor.apply();

        res.success = true;
        res.message = summary.toString();
        return res;
    }

    /** Parse an ATAK preference XML and return entries from the {@code cot_streams} block. */
    private static Map<String, String> parseCotStreamsPref(String xml) throws XmlPullParserException, IOException {
        Map<String, String> values = new HashMap<>();
        XmlPullParser parser = Xml.newPullParser();
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
        parser.setInput(new StringReader(xml));

        boolean inCotStreams = false;
        String currentKey = null;
        StringBuilder textBuf = new StringBuilder();

        int event = parser.getEventType();
        while (event != XmlPullParser.END_DOCUMENT) {
            switch (event) {
                case XmlPullParser.START_TAG:
                    String tag = parser.getName();
                    if ("preference".equals(tag)) {
                        inCotStreams = "cot_streams".equals(parser.getAttributeValue(null, "name"));
                    } else if ("entry".equals(tag) && inCotStreams) {
                        currentKey = parser.getAttributeValue(null, "key");
                        textBuf.setLength(0);
                    }
                    break;
                case XmlPullParser.TEXT:
                    if (currentKey != null) textBuf.append(parser.getText());
                    break;
                case XmlPullParser.END_TAG:
                    String endTag = parser.getName();
                    if ("entry".equals(endTag) && currentKey != null) {
                        values.put(currentKey, textBuf.toString().trim());
                        currentKey = null;
                        textBuf.setLength(0);
                    } else if ("preference".equals(endTag)) {
                        inCotStreams = false;
                    }
                    break;
            }
            event = parser.next();
        }
        return values;
    }

    /**
     * Test a TCP (and, if {@code ssl}, TLS) connection to {@code host:port}. Network call —
     * must run off the main thread.
     */
    public static ConnectionTestResult testConnection(Result res, int timeoutMs) {
        ConnectionTestResult ct = new ConnectionTestResult();
        if (res == null || res.host == null || res.port <= 0) {
            ct.details = "no host/port to test";
            return ct;
        }
        // Plain TCP reachability.
        try (Socket s = new Socket()) {
            s.connect(new InetSocketAddress(res.host, res.port), timeoutMs);
            ct.tcpOk = true;
        } catch (Exception e) {
            ct.details = "TCP connect failed: " + e.getMessage();
            return ct;
        }

        if (!res.ssl) {
            ct.tlsOk = true;
            ct.details = "TCP reachable";
            return ct;
        }

        // TLS handshake with client cert + trust store, if present.
        try {
            KeyManagerFactory kmf = null;
            if (res.clientCertPath != null) {
                KeyStore clientKs = KeyStore.getInstance("PKCS12");
                try (FileInputStream fis = new FileInputStream(res.clientCertPath)) {
                    clientKs.load(fis, res.clientCertPassword == null ? new char[0] : res.clientCertPassword.toCharArray());
                }
                kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
                kmf.init(clientKs, res.clientCertPassword == null ? new char[0] : res.clientCertPassword.toCharArray());
            }

            TrustManagerFactory tmf = null;
            if (res.trustStorePath != null) {
                KeyStore trustKs = KeyStore.getInstance("PKCS12");
                try (FileInputStream fis = new FileInputStream(res.trustStorePath)) {
                    trustKs.load(fis, res.trustStorePassword == null ? new char[0] : res.trustStorePassword.toCharArray());
                }
                tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
                tmf.init(trustKs);
            }

            SSLContext ctx = SSLContext.getInstance("TLS");
            ctx.init(kmf == null ? null : kmf.getKeyManagers(),
                    tmf == null ? null : tmf.getTrustManagers(),
                    new SecureRandom());

            SSLSocketFactory factory = ctx.getSocketFactory();
            try (SSLSocket ssl = (SSLSocket) factory.createSocket()) {
                ssl.connect(new InetSocketAddress(res.host, res.port), timeoutMs);
                ssl.setSoTimeout(timeoutMs);
                ssl.startHandshake();
                ct.tlsOk = true;
                ct.details = "TLS handshake OK (" + ssl.getSession().getProtocol() + ")";
            }
        } catch (Exception e) {
            ct.details = "TLS handshake failed: " + e.getMessage();
        }
        return ct;
    }
}

File: ios/Sentinel/Services/TakDataPackageImporter.swift
//
//  TakDataPackageImporter.swift
//  Mirrors the Android TakDataPackageImporter — reads an ATAK / TAK Server data-package .zip,
//  extracts the `.pref` XML + any `.p12` cert files, parses the `cot_streams` preference block,
//  and applies the matching keys to UserDefaults (same key names as the Android side so the
//  same .zip works on both apps).
//
//  Zip layout (per ATAK spec):
//    *.pref     — XML preference file with a <preference name="cot_streams"> block
//    *.p12      — client cert (referenced by certificateLocation0) + trust store (caLocation0)
//
//  After import:
//    • atak_address      = host  (from connectString0 = "host:port:protocol")
//    • atak_port         = port
//    • atak_ssl          = (protocol ∈ {ssl, tls, https})
//    • atak_send_cot     = true (if enabled0 == "true")
//    • Files written under Documents/TAK/<basename>; absolute paths stored in
//      trust_store_certificate / client_certificate prefs.
//

import Foundation
import ZIPFoundation

struct TakImportResult {
    var success: Bool = false
    var message: String = ""
    var host: String?
    var port: Int = 0
    var ssl: Bool = false
    var trustStorePath: URL?
    var trustStorePassword: String?
    var clientCertPath: URL?
    var clientCertPassword: String?
    var description: String?
}

struct TakConnectionTest {
    var tcpOk: Bool = false
    var tlsOk: Bool = false
    var details: String = ""
}

enum TakImporter {

    /// Imports a TAK data package from a local file URL. The file may be a `file://`-scheme URL
    /// from the SwiftUI `.fileImporter`. Returns a result describing what was applied.
    @MainActor
    static func importPackage(at zipURL: URL) async -> TakImportResult {
        var result = TakImportResult()

        // 1. Open the zip.
        let archive: Archive
        do {
            archive = try Archive(url: zipURL, accessMode: .read)
        } catch {
            result.message = "Could not open zip: \(error.localizedDescription)"
            return result
        }

        // 2. Walk entries: pull the .pref XML into memory; extract .p12/.pem/.jks to sandbox.
        let takDir: URL
        do {
            takDir = try takDirectory()
        } catch {
            result.message = "Could not access sandbox: \(error.localizedDescription)"
            return result
        }

        var prefXML: String?
        var extracted: [String: URL] = [:]   // basename → on-disk URL

        for entry in archive {
            guard entry.type == .file else { continue }
            let normalised = entry.path.replacingOccurrences(of: "\\", with: "/")
            let basename = (normalised as NSString).lastPathComponent
            let lower = basename.lowercased()

            if lower.hasSuffix(".pref") {
                var buf = Data()
                _ = try? archive.extract(entry) { data in buf.append(data) }
                prefXML = String(data: buf, encoding: .utf8)
            } else if lower.hasSuffix(".p12") || lower.hasSuffix(".pem") || lower.hasSuffix(".jks") {
                let dest = takDir.appendingPathComponent(basename)
                try? FileManager.default.removeItem(at: dest)
                _ = try? archive.extract(entry, to: dest)
                extracted[basename] = dest
            }
        }

        guard let xml = prefXML else {
            result.message = "No .pref file found inside the zip."
            return result
        }

        // 3. Parse cot_streams.
        let values = parseCotStreams(xml: xml)
        guard let connect = values["connectString0"] else {
            result.message = "The .pref file is missing connectString0."
            return result
        }

        // 4. Apply to UserDefaults.
        let defaults = UserDefaults.standard
        var summary = ""

        if let desc = values["description0"] { result.description = desc; summary += "Imported: \(desc)\n\n" }

        let parts = connect.split(separator: ":").map(String.init)
        if parts.indices.contains(0) {
            result.host = parts[0]
            defaults.set(parts[0], forKey: Pref.atakServerAddress)
            summary += "Server: \(parts[0])\n"
        }
        if parts.indices.contains(1) {
            result.port = Int(parts[1].trimmingCharacters(in: .whitespaces)) ?? 0
            defaults.set(parts[1], forKey: Pref.atakServerPort)
            summary += "Port: \(parts[1])\n"
        }
        if parts.indices.contains(2) {
            let proto = parts[2].lowercased()
            let isSecure = (proto == "ssl" || proto == "https" || proto == "tls")
            result.ssl = isSecure
            defaults.set(isSecure, forKey: Pref.atakSSL)
            summary += "Protocol: \(parts[2])\n"
        }

        if values["enabled0"]?.lowercased() == "true" {
            defaults.set(true, forKey: Pref.atakSendCoT)
            summary += "CoT streaming: enabled\n"
        }

        if let caLoc = values["caLocation0"] {
            let basename = ((caLoc as NSString).replacingOccurrences(of: "\\", with: "/") as NSString).lastPathComponent
            if let url = extracted[basename] {
                result.trustStorePath = url
                defaults.set(url.path, forKey: Pref.atakTrustStorePath)
                summary += "Trust store: \(basename)\n"
            }
        }
        if let caPwd = values["caPassword0"] {
            result.trustStorePassword = caPwd
            defaults.set(caPwd, forKey: Pref.atakTrustStorePassword)
        }
        if let certLoc = values["certificateLocation0"] {
            let basename = ((certLoc as NSString).replacingOccurrences(of: "\\", with: "/") as NSString).lastPathComponent
            if let url = extracted[basename] {
                result.clientCertPath = url
                defaults.set(url.path, forKey: Pref.atakClientCertPath)
                summary += "Client cert: \(basename)\n"
            }
        }
        if let clientPwd = values["clientPassword0"] {
            result.clientCertPassword = clientPwd
            defaults.set(clientPwd, forKey: Pref.atakClientCertPassword)
        }

        result.success = true
        result.message = summary.isEmpty ? "Imported." : summary
        return result
    }

    // MARK: - XML parsing (cot_streams entries)

    /// XMLParser-based equivalent of the Android pull-parser walk: find
    /// `<preference name="cot_streams">…</preference>`, harvest every `<entry key="…">value</entry>`.
    private static func parseCotStreams(xml: String) -> [String: String] {
        let parser = XMLParser(data: Data(xml.utf8))
        let delegate = CotStreamsParser()
        parser.delegate = delegate
        parser.parse()
        return delegate.values
    }

    // MARK: - Helpers

    private static func takDirectory() throws -> URL {
        let docs = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask,
                                               appropriateFor: nil, create: true)
        let dir = docs.appendingPathComponent("TAK", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }
}

// XMLParserDelegate isolates per-import state.
private final class CotStreamsParser: NSObject, XMLParserDelegate {
    var values: [String: String] = [:]
    private var inCotStreams = false
    private var currentKey: String?
    private var buffer = ""

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?,
                qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
        switch elementName {
        case "preference":
            inCotStreams = (attributeDict["name"] == "cot_streams")
        case "entry":
            if inCotStreams {
                currentKey = attributeDict["key"]
                buffer = ""
            }
        default: break
        }
    }
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if currentKey != nil { buffer += string }
    }
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?,
                qualifiedName qName: String?) {
        if elementName == "entry", let key = currentKey {
            values[key] = buffer.trimmingCharacters(in: .whitespacesAndNewlines)
            currentKey = nil
            buffer = ""
        } else if elementName == "preference" {
            inCotStreams = false
        }
    }
}
